yarn-project/p2p/src/client/p2p_client.ts

```diff
@@ -558,6 +602,36 @@ export class P2PClient extends WithTracer implements P2P {
     await this.startServiceIfSynched();
   }
 
+  /**
+   * Updates the tx pool after a chain prune.
+   * @param latestBlock - The block number the chain was pruned to.
+   */
+  private async handlePruneL2Blocks(latestBlock: number): Promise<void> {
+    const txsToDelete: TxHash[] = [];
+    for (const tx of this.txPool.getAllTxs()) {
+      // every tx that's been generated against a block that has now been pruned is no longer valid
+      // NOTE (alexg): I think this check against block hash instead of block number?
+      if (tx.data.constants.globalVariables.blockNumber.toNumber() > latestBlock) {
+        txsToDelete.push(tx.getTxHash());
+      }
+    }
+
+    this.log.info(
+      `Detected chain prune. Removing invalid txs count=${
+        txsToDelete.length
+      } newLatestBlock=${latestBlock} previousLatestBlock=${this.getSyncedLatestBlockNum()}`,
+    );
+
+    // delete invalid txs (both pending and mined)
+    await this.txPool.deleteTxs(txsToDelete);
+    // everything left in the mined set was built against a block on the proven chain so its still valid
+    // move back to pending set
+    await this.txPool.markMinedAsPending(this.txPool.getMinedTxHashes());
```

Comment:
Why are we moving **all** mined txs to pending? We should only move the ones that were mined in a block after `latestBlock`, right?

Reply:
My logic is: the set of mined transactions at the point of a chain prune will be made up of two subsets: the set of all invalid transactions (ie. those that were built against a block that no longer exists) and everything else.

The "everything else" set will contain all transactions that are still good (i.e. built from a finalized block) but were mined in a pruned block so should be safe to re-add to pending as every other transaction was either removed or finalised.

WDYT?

Reply:
I think it works only if `keepProvenTxsFor` is set to zero. Otherwise, the set of mined txs will include some txs that are on the proven chain, which should not get pruned, and so should not be moved to pending. But to be fair, we had only included `keepProvenTxsFor` to account for scenarios where a prover may want to prove something already proven (ie provernet v1), but it's no longer relevant.

Reply:
Note this may change when we add support for L1 reorgs. We'll want to clear txs from the p2p pool once they are *finalized* (in the sense of them being in a block that is proven and L1 finalized) and not just _proven_. Then this line would have to change to sth like `markMinedAsPending(txPool.getMinedButNotProvenTxHashes())`.

Reply:
Ok, sounds like we can go ahead and merge and deal with that condition later.

Reply:
Changed this in the last commit to track which block mined a tx. This should also cover the case where we hold onto txs for longer than the prove chain (`this.keepProvenTxsFor`)